pkg/virtctl/create/vm/vm.go

```diff
@@ -556,6 +603,119 @@ func (c *createVM) getInferFromVolume(vm *v1.VirtualMachine) (string, error) {
 	return vm.Spec.Template.Spec.Volumes[0].Name, nil
 }
 
+func (c *createVM) cloudInitConfig(cmd *cobra.Command, vm *v1.VirtualMachine) error {
+	if !cmd.Flags().Changed(UserFlag) &&
+		!cmd.Flags().Changed(GenPasswordFlag) &&
+		!cmd.Flags().Changed(PasswordFileFlag) &&
+		!cmd.Flags().Changed(SSHKeyFlag) &&
+		!cmd.Flags().Changed(GAManageSSHFlag) &&
+		!cmd.Flags().Changed(CloudInitSourceTypeFlag) &&
+		!cmd.Flags().Changed(CloudInitUserDataFlag) &&
+		!cmd.Flags().Changed(CloudInitNetworkDataFlag) {
+		return nil
+	}
+
+	// Make sure cloudInitDisk does not already exist
+	if vol := volumeExists(vm, CloudInitDisk); vol != nil {
+		return fmt.Errorf(VolumeExistsErrorFmt, CloudInitDisk)
+	}
+
+	var src v1.VolumeSource
+	var err error
+	switch c.cloudInitSourceType {
+	case string(cloudinit.DataSourceNoCloud):
+		src, err = c.noCloudVolumeSource()
+	case string(cloudinit.DataSourceConfigDrive):
+		src, err = c.configDriveVolumeSource()
+	default:
+		return params.FlagErr(CloudInitSourceTypeFlag, "invalid cloud init datasource type \"%s\", supported values are: %s, %s", c.cloudInitSourceType, string(cloudinit.DataSourceNoCloud), string(cloudinit.DataSourceConfigDrive))
+	}
+	if err != nil {
+		return err
+	}
+
+	vm.Spec.Template.Spec.Volumes = append(vm.Spec.Template.Spec.Volumes, v1.Volume{
+		Name:         CloudInitDisk,
+		VolumeSource: src,
+	})
+
+	return nil
+}
+
+func (c *createVM) noCloudVolumeSource() (v1.VolumeSource, error) {
+	src := &v1.CloudInitNoCloudSource{}
+	if c.cloudInitNetworkData != "" {
+		src.NetworkDataBase64 = c.cloudInitNetworkData
+	}
+	if c.cloudInitUserData != "" {
+		src.UserDataBase64 = c.cloudInitUserData
+	} else {
+		config, err := c.buildCloudInitConfig()
+		if err != nil {
+			return v1.VolumeSource{}, err
+		}
+		src.UserData = config
+	}
+	return v1.VolumeSource{
+		CloudInitNoCloud: src,
+	}, nil
+}
+
+func (c *createVM) configDriveVolumeSource() (v1.VolumeSource, error) {
+	src := &v1.CloudInitConfigDriveSource{}
+	if c.cloudInitNetworkData != "" {
+		src.NetworkDataBase64 = c.cloudInitNetworkData
+	}
+	if c.cloudInitUserData != "" {
+		src.UserDataBase64 = c.cloudInitUserData
+	} else {
+		config, err := c.buildCloudInitConfig()
+		if err != nil {
+			return v1.VolumeSource{}, err
+		}
+		src.UserData = config
+	}
+	return v1.VolumeSource{
+		CloudInitConfigDrive: src,
+	}, nil
+}
+
+func (c *createVM) buildCloudInitConfig() (string, error) {
+	config := "#cloud-config"
+
+	if c.user != "" {
+		config += "\nuser: " + c.user
+	}
+
+	password := ""
+	if c.genPassword {
+		password = rand.String(12)
+	}
+	if c.passwordFile != "" {
+		data, err := os.ReadFile(c.passwordFile)
+		if err != nil {
+			return "", params.FlagErr(PasswordFileFlag, "%w", err)
+		}
+		password = strings.TrimSpace(string(data))
+	}
```

Comment:
I'm honestly not sure about dumping raw passwords in the VM definitions like this.

Anyone with read access on the VMs is going to see this right?

Reply:
Right, given our medium for storing and transmitting this config is a VirtualMachine object in k8s I really don't think we should directly enable this through `virtctl`.

Reply:
Yup that's a valid comprimise for now given it's already possible for users to do this outside of `virtctl`.